File: 1236-n-th-tribonacci-number/n-th-tribonacci-number.py
class Solution(object):
    def tribonacci(self, n):
        """
        :type n: int
        :rtype: int
        """
        memo = {}
        def DP(index):
            if index in memo:
                return memo[index]

            if index == 0:
                return 0

            if 0<index<=2:
                return 1

            memo[index] = DP(index-3) + DP(index -2) + DP(index -1)
            return memo[index]

        return DP(n)

File: 1236-n-th-tribonacci-number/n-th-tribonacci-number.cpp
class Solution {
public:
    int DP(int n, std::unordered_map<int,int>& memo)
    {
        // base cases
        if(n == 0)
        {
            return 0;
        }

        if(n<=2)
        {
            return 1;
        }

        // check precomputed
        if(memo.find(n) == memo.end())
        {
            memo[n] = DP(n-3,memo) + DP(n-2,memo) + DP(n-1,memo);
        }

        return memo[n];
    }
    int tribonacci(int n) {

        std::unordered_map<int,int> memo;
        return DP(n,memo);
        
    }
};